File: app_java/src/main/java/tech/thdev/android_mvp_sample/view/main/presenter/MainPresenter.java
package tech.thdev.android_mvp_sample.view.main.presenter;

import android.content.Context;

import java.util.ArrayList;

import tech.thdev.android_mvp_sample.adapter.contract.ImageAdapterContract;
import tech.thdev.android_mvp_sample.data.ImageItem;
import tech.thdev.android_mvp_sample.data.source.image.SampleImageRepository;
import tech.thdev.android_mvp_sample.data.source.image.SampleImageSource;
import tech.thdev.android_mvp_sample.listener.OnItemClickListener;

/**
 * Created by tae-hwan on 12/22/16.
 */

public class MainPresenter implements MainContract.Presenter, OnItemClickListener {

    private MainContract.View view;

    private ImageAdapterContract.Model adapterModel;
    private ImageAdapterContract.View adapterView;

    private SampleImageRepository sampleImageData;

    @Override
    public void attachView(MainContract.View view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        view = null;
    }

    @Override
    public void setSampleImageData(SampleImageRepository sampleImageData) {
        this.sampleImageData = sampleImageData;
    }

    @Override
    public void loadItems(Context context, final boolean isClear) {
        sampleImageData.getImages(context, 10, new SampleImageSource.LoadImageCallback() {
            @Override
            public void onImageLoaded(ArrayList<ImageItem> list) {
                if (list != null) {
                    if (isClear) {
                        adapterModel.clearItem();
                    }
                    adapterModel.addItems(list);
                    adapterView.notifyAdapter();
                }
            }
        });


    }

    @Override
    public void setImageAdapterModel(ImageAdapterContract.Model adapterModel) {
        this.adapterModel = adapterModel;
    }

    @Override
    public void setImageAdapterView(ImageAdapterContract.View adapterView) {
        this.adapterView = adapterView;

        this.adapterView.setOnClickListener(this);
    }

    @Override
    public void onItemClick(int position) {
        ImageItem imageItem = adapterModel.getItem(position);
        view.showToast(imageItem.getTitle());
    }
}

File: app_kotlin/src/main/java/tech/thdev/app_kotlin/view/main/presenter/MainPresenter.kt
package tech.thdev.app_kotlin.view.main.presenter

import android.content.Context
import tech.thdev.app_kotlin.adapter.contract.ImageAdapterContract
import tech.thdev.app_kotlin.data.ImageItem
import tech.thdev.app_kotlin.data.source.image.SampleImageRepository
import tech.thdev.app_kotlin.data.source.image.SampleImageSource
import java.util.*

/**
 * Created by tae-hwan on 12/23/16.
 */
class MainPresenter(
    private val view: MainContract.View,
    private val imageData: SampleImageRepository,
    private val adapterModel: ImageAdapterContract.Model,
    private val adapterView: ImageAdapterContract.View
) : MainContract.Presenter {

    init {
        adapterView.onClickFunc = {
            onClickListener(it)
        }
    }

    override fun loadItems(context: Context, isClear: Boolean) {
        imageData.getImages(context, 10, object : SampleImageSource.LoadImageCallback {
            override fun onLoadImages(list: ArrayList<ImageItem>) {
                if (isClear) {
                    adapterModel.clearItem()
                }

                adapterModel.addItems(list)
                adapterView.notifyAdapter()
            }
        })
    }

    private fun onClickListener(position: Int) {
        adapterModel.getItem(position).let {
            view.showToast(it.title)
        }
    }
}